File: custom_components/local_mdm/policy.py
"""Policy document model and the network-safety guard.

This module has no Home Assistant imports so it can be unit tested and reused
by tooling. The guard is the control-plane half of a two-sided rule; the DPC
refuses the same keys. docs/security.md explains why both sides check.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Final

from .const import (
    APP_MODE_ALLOWLIST,
    APP_MODE_OPEN,
    APP_MODES,
    POLICY_ALLOWED_PACKAGES,
    POLICY_APP_MODE,
    POLICY_FLAGS,
    POLICY_KIOSK_MODE,
    POLICY_KIOSK_PACKAGES,
)

# Android UserManager restriction names and policy keys that can sever the
# only management path (Wi-Fi, adb). Never accepted, never forwarded.
FORBIDDEN_KEYS: Final[frozenset[str]] = frozenset(
    {
        "no_config_wifi",
        "no_config_tethering",
        "no_network_reset",
        "no_config_mobile_networks",
        "no_airplane_mode",
        "no_debugging_features",
        "no_config_private_dns",
        "no_config_vpn",
        "wifi_disabled",
        "network_settings_blocked",
        "adb_disabled",
        "user_restrictions",
        "global_settings",
        "secure_settings",
    }
)

DPC_PACKAGE: Final = "com.trooperthorn.localmdm"


class UnsafePolicyError(ValueError):
    """Raised for a policy that could remove the management path."""


class InvalidPolicyError(ValueError):
    """Raised for a policy with unknown keys or wrong value types."""


def default_policy() -> dict[str, Any]:
    """Return the least restrictive policy: every flag off, no kiosk packages."""
    policy: dict[str, Any] = dict.fromkeys(POLICY_FLAGS, False)
    policy[POLICY_KIOSK_PACKAGES] = []
    policy[POLICY_APP_MODE] = APP_MODE_OPEN
    policy[POLICY_ALLOWED_PACKAGES] = []
    return policy


def validate_policy(policy: Mapping[str, Any]) -> dict[str, Any]:
    """Return a normalized copy of ``policy`` or raise.

    The forbidden-key check runs before the unknown-key check so a forbidden
    key is reported as unsafe rather than merely unknown.
    """
    forbidden = FORBIDDEN_KEYS.intersection(policy)
    if forbidden:
        raise UnsafePolicyError(
            f"Policy keys {sorted(forbidden)} can sever the management path and are refused"
        )

    allowed = set(POLICY_FLAGS) | {POLICY_KIOSK_PACKAGES, POLICY_APP_MODE, POLICY_ALLOWED_PACKAGES}
    unknown = set(policy) - allowed
    if unknown:
        raise InvalidPolicyError(f"Unknown policy keys: {sorted(unknown)}")

    normalized = default_policy()
    for key in POLICY_FLAGS:
        if key in policy:
            value = policy[key]
            if not isinstance(value, bool):
                raise InvalidPolicyError(f"{key} must be a boolean, got {value!r}")
            normalized[key] = value

    for key in (POLICY_KIOSK_PACKAGES, POLICY_ALLOWED_PACKAGES):
        packages = policy.get(key, [])
        if not isinstance(packages, list) or not all(isinstance(p, str) and p for p in packages):
            raise InvalidPolicyError(f"{key} must be a list of package names")
        normalized[key] = sorted(set(packages) - {DPC_PACKAGE})

    mode = policy.get(POLICY_APP_MODE, APP_MODE_OPEN)
    if mode not in APP_MODES:
        raise InvalidPolicyError(
            f"{POLICY_APP_MODE} must be one of {list(APP_MODES)}, got {mode!r}"
        )
    normalized[POLICY_APP_MODE] = mode
    if mode == APP_MODE_ALLOWLIST:
        # The kiosk target must stay openable, whichever mode is active.
        normalized[POLICY_ALLOWED_PACKAGES] = sorted(
            set(normalized[POLICY_ALLOWED_PACKAGES]) | set(normalized[POLICY_KIOSK_PACKAGES])
        )
        if not normalized[POLICY_ALLOWED_PACKAGES]:
            raise UnsafePolicyError(
                "app_mode allowlist requires at least one package in allowed_packages; "
                "suspending every application leaves the operator nothing to open"
            )

    if normalized[POLICY_KIOSK_MODE] and not normalized[POLICY_KIOSK_PACKAGES]:
        raise UnsafePolicyError(
            "kiosk_mode requires at least one package in kiosk_packages; "
            "locking the tablet to the DPC alone leaves the operator no application"
        )
    return normalized

File: android/app/src/main/java/com/trooperthorn/localmdm/Policy.kt
package com.trooperthorn.localmdm

import org.json.JSONArray
import org.json.JSONObject

/**
 * The policy document shared with the Home Assistant integration.
 * Keys and the forbidden set must match custom_components/local_mdm/policy.py;
 * docs/protocol.md is the contract both sides are written against.
 */
data class Policy(
    val kioskMode: Boolean = false,
    val cameraDisabled: Boolean = false,
    val screenCaptureDisabled: Boolean = false,
    val statusBarDisabled: Boolean = false,
    val installAppsBlocked: Boolean = false,
    val uninstallAppsBlocked: Boolean = false,
    val usbFileTransferBlocked: Boolean = false,
    val adjustVolumeBlocked: Boolean = false,
    val safeBootBlocked: Boolean = false,
    val factoryResetBlocked: Boolean = false,
    val autoOsUpdates: Boolean = false,
    val stayAwakeOnPower: Boolean = false,
    val kioskPackages: List<String> = emptyList(),
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put(KEY_KIOSK_MODE, kioskMode)
        put(KEY_CAMERA_DISABLED, cameraDisabled)
        put(KEY_SCREEN_CAPTURE_DISABLED, screenCaptureDisabled)
        put(KEY_STATUS_BAR_DISABLED, statusBarDisabled)
        put(KEY_INSTALL_APPS_BLOCKED, installAppsBlocked)
        put(KEY_UNINSTALL_APPS_BLOCKED, uninstallAppsBlocked)
        put(KEY_USB_FILE_TRANSFER_BLOCKED, usbFileTransferBlocked)
        put(KEY_ADJUST_VOLUME_BLOCKED, adjustVolumeBlocked)
        put(KEY_SAFE_BOOT_BLOCKED, safeBootBlocked)
        put(KEY_FACTORY_RESET_BLOCKED, factoryResetBlocked)
        put(KEY_AUTO_OS_UPDATES, autoOsUpdates)
        put(KEY_STAY_AWAKE_ON_POWER, stayAwakeOnPower)
        put(KEY_KIOSK_PACKAGES, JSONArray(kioskPackages))
    }

    companion object {
        const val KEY_KIOSK_MODE = "kiosk_mode"
        const val KEY_CAMERA_DISABLED = "camera_disabled"
        const val KEY_SCREEN_CAPTURE_DISABLED = "screen_capture_disabled"
        const val KEY_STATUS_BAR_DISABLED = "status_bar_disabled"
        const val KEY_INSTALL_APPS_BLOCKED = "install_apps_blocked"
        const val KEY_UNINSTALL_APPS_BLOCKED = "uninstall_apps_blocked"
        const val KEY_USB_FILE_TRANSFER_BLOCKED = "usb_file_transfer_blocked"
        const val KEY_ADJUST_VOLUME_BLOCKED = "adjust_volume_blocked"
        const val KEY_SAFE_BOOT_BLOCKED = "safe_boot_blocked"
        const val KEY_FACTORY_RESET_BLOCKED = "factory_reset_blocked"
        const val KEY_AUTO_OS_UPDATES = "auto_os_updates"
        const val KEY_STAY_AWAKE_ON_POWER = "stay_awake_on_power"
        const val KEY_KIOSK_PACKAGES = "kiosk_packages"

        val FLAG_KEYS = listOf(
            KEY_KIOSK_MODE, KEY_CAMERA_DISABLED, KEY_SCREEN_CAPTURE_DISABLED,
            KEY_STATUS_BAR_DISABLED, KEY_INSTALL_APPS_BLOCKED, KEY_UNINSTALL_APPS_BLOCKED,
            KEY_USB_FILE_TRANSFER_BLOCKED, KEY_ADJUST_VOLUME_BLOCKED, KEY_SAFE_BOOT_BLOCKED,
            KEY_FACTORY_RESET_BLOCKED, KEY_AUTO_OS_UPDATES, KEY_STAY_AWAKE_ON_POWER,
        )

        /** Keys that could sever Wi-Fi or adb, the only recovery paths. Refused on sight. */
        val FORBIDDEN_KEYS = setOf(
            "no_config_wifi", "no_config_tethering", "no_network_reset",
            "no_config_mobile_networks", "no_airplane_mode", "no_debugging_features",
            "no_config_private_dns", "no_config_vpn", "wifi_disabled",
            "network_settings_blocked", "adb_disabled", "user_restrictions",
            "global_settings", "secure_settings",
        )

        class UnsafePolicy(message: String) : IllegalArgumentException(message)
        class InvalidPolicy(message: String) : IllegalArgumentException(message)

        fun fromJson(json: JSONObject, selfPackage: String): Policy {
            val keys = json.keys().asSequence().toSet()
            val forbidden = keys.intersect(FORBIDDEN_KEYS)
            if (forbidden.isNotEmpty()) {
                throw UnsafePolicy("Refusing keys that can sever the management path: $forbidden")
            }
            val unknown = keys - FLAG_KEYS.toSet() - KEY_KIOSK_PACKAGES
            if (unknown.isNotEmpty()) throw InvalidPolicy("Unknown policy keys: $unknown")

            fun flag(key: String): Boolean {
                if (!json.has(key)) return false
                val value = json.get(key)
                if (value !is Boolean) throw InvalidPolicy("$key must be a boolean")
                return value
            }

            val packages = mutableSetOf<String>()
            json.optJSONArray(KEY_KIOSK_PACKAGES)?.let { array ->
                for (i in 0 until array.length()) {
                    val name = array.optString(i, "")
                    if (name.isEmpty()) throw InvalidPolicy("kiosk_packages entries must be package names")
                    if (name != selfPackage) packages.add(name)
                }
            }
            val policy = Policy(
                kioskMode = flag(KEY_KIOSK_MODE),
                cameraDisabled = flag(KEY_CAMERA_DISABLED),
                screenCaptureDisabled = flag(KEY_SCREEN_CAPTURE_DISABLED),
                statusBarDisabled = flag(KEY_STATUS_BAR_DISABLED),
                installAppsBlocked = flag(KEY_INSTALL_APPS_BLOCKED),
                uninstallAppsBlocked = flag(KEY_UNINSTALL_APPS_BLOCKED),
                usbFileTransferBlocked = flag(KEY_USB_FILE_TRANSFER_BLOCKED),
                adjustVolumeBlocked = flag(KEY_ADJUST_VOLUME_BLOCKED),
                safeBootBlocked = flag(KEY_SAFE_BOOT_BLOCKED),
                factoryResetBlocked = flag(KEY_FACTORY_RESET_BLOCKED),
                autoOsUpdates = flag(KEY_AUTO_OS_UPDATES),
                stayAwakeOnPower = flag(KEY_STAY_AWAKE_ON_POWER),
                kioskPackages = packages.sorted(),
            )
            if (policy.kioskMode && policy.kioskPackages.isEmpty()) {
                throw UnsafePolicy("kiosk_mode requires at least one package in kiosk_packages")
            }
            return policy
        }
    }
}
